### ISSUE TITLE
What's the usage of '__constant__' in FPN module?

### ISSUE THREAD

**Issue by feiyuhuahuo** [dbolya/yolact#127]
https://github.com/dbolya/yolact/blob/29e809e563d0c3e6c47c5c6716b8cc81686a2b24/yolact.py#L299-L300
I'm not familiar with `__constant__`. As far as I know, it's for some utility  with `torch.jit.ScriptModule`. And I found because of this, all the `nn.Conv2d` in FPN module were not initialized by `xavier`. Would this has some effect on the final performance?

**Comment by dbolya** [comment]
Oh weird, you found this empirically? `__constants__` is used to (a) lock in constants like booleans and strings so JIT can assume they won't change (and thus can bake them into the architecture), and (b) tell JIT that the lengths of ModuleLists will not change so it can unroll those ModuleLists in loops and such. It should have any effect as to whether or not you can modify those weights for initialization.

So if you found that they're not getting initialized by init_weights empirically, can you let me know how you went about it? That would be a bug, and I'm not sure what the impact would be, so hopefully  I can fix it.

**Comment by feiyuhuahuo** [comment]
I modified the code in `init_weights` in `yolact.py`.
![3](https://user-images.githubusercontent.com/32631344/63986816-40de5400-cb08-11e9-8355-42dd647a59ba.png)
Then the result is:
![1](https://user-images.githubusercontent.com/32631344/63986841-55bae780-cb08-11e9-974d-8170dd9a4842.png)
Then I commented the `__constants__` and the decorator `@script_method_wrapper` in FPN module, the result is:
![2](https://user-images.githubusercontent.com/32631344/63987046-37092080-cb09-11e9-9917-33c661596f9c.png)

**Comment by dbolya** [comment]
Yup, that's definitely a bug. And uhg I've done some testing and this is going to be difficult to fix.

Basically, what happened, is when I used those parameters in a script module as constants, pytorch replaced them with "WeakScriptModuleProxy" objects, which copy all the parameters over from the original module. Now, when I test the type of this new module, it only returns "WeakScriptModuleProxy" and not the original Conv.

Pytorch attempts to store a weak reference to the original module inside the proxy (using `module.__dict__['_original']`, but the reference is dead by the time initialization rolls around so I can't read its type.

On the other hand, the newest version of Pytorch (1.2 and maybe 1.1) adds a `module.__dict__['_original_type']` field that stores the original type with no reference. So if I stopped supporting Pytorch 1.0.1, I could use that feature to test for the type. But I really don't want to do that...

I'll see if I can figure something out, and I'll push a fix if I can.

**Comment by dbolya** [comment]
Ok fixed with the latest commit. Let me know if this works for you--I had to do a really hacky solution taking the constants in the WeakScriptModuleProxy module and comparing them with a Conv2d layer's constants (using the constants as a "hash" of the module type). Those constants are `{'groups', 'dilation', 'padding', 'stride', 'bias'}`, so if another type of layer has the exact same constants there might be some issues (though I doubt this would happen).

Here's what was initialized before:
```
proto_net.0
proto_net.2
proto_net.4
proto_net.8
proto_net.10
prediction_layers.0.upfeature.0
prediction_layers.0.bbox_layer
prediction_layers.0.conf_layer
prediction_layers.0.mask_layer
semantic_seg_conv
```

And after:
```
proto_net.0
proto_net.2
proto_net.4
proto_net.8
proto_net.10
fpn.lat_layers.0
fpn.lat_layers.1
fpn.lat_layers.2
fpn.pred_layers.0
fpn.pred_layers.1
fpn.pred_layers.2
fpn.downsample_layers.0
fpn.downsample_layers.1
prediction_layers.0.upfeature.0
prediction_layers.0.bbox_layer
prediction_layers.0.conf_layer
prediction_layers.0.mask_layer
semantic_seg_conv
```

All the fpn layers are now there as expected.

**Comment by feiyuhuahuo** [comment]
Yes, the new fix works for me. And I'd like to ask if I just let the FPN class inherit `nn.Module` and delete `__constants__`. Then I train the network. Will this affect the training results?

**Comment by dbolya** [comment]
@feiyuhuahuo 
Nah, it'll only affect speed. Also, if you've trained like that, you can add the ScriptModule and __constants__ back and there would be (hopefully) no change in evaluation performance, just evaluation speed.

Though, I'm kind of underwhelmed by torch's JIT functionality. It doesn't seem to be much faster for things that it should be much faster for (e.g., the ubiquitous conv + relu + batch norm).
